File: ios/Classes/strategies/AudioVideoPermissionStrategy.m
//
// Created by Razvan Lung(long1eu) on 2019-02-15.
// Copyright (c) 2019 The Chromium Authors. All rights reserved.
//

#import "AudioVideoPermissionStrategy.h"

#if PERMISSION_CAMERA | PERMISSION_MICROPHONE

@implementation AudioVideoPermissionStrategy

- (PermissionStatus)checkPermissionStatus:(PermissionGroup)permission {
    if (permission == PermissionGroupCamera) {
        #if PERMISSION_CAMERA
        return [AudioVideoPermissionStrategy permissionStatus:AVMediaTypeVideo];
        #else
        return PermissionStatusUnknown;
        #endif
    } else if (permission == PermissionGroupMicrophone) {
        #if PERMISSION_MICROPHONE
        return [AudioVideoPermissionStrategy permissionStatus:AVMediaTypeAudio];
        #else
        return PermissionStatusUnknown;
        #endif
    }
    return PermissionStatusUnknown;
}

- (ServiceStatus)checkServiceStatus:(PermissionGroup)permission {
    return ServiceStatusNotApplicable;
}

- (void)requestPermission:(PermissionGroup)permission completionHandler:(PermissionStatusHandler)completionHandler {
    PermissionStatus status = [self checkPermissionStatus:permission];

    if (status != PermissionStatusUnknown) {
        completionHandler(status);
        return;
    }

    AVMediaType mediaType;

    if (permission == PermissionGroupCamera) {
        #if PERMISSION_CAMERA
        mediaType = AVMediaTypeVideo;
        #else
        completionHandler(PermissionStatusUnknown);
        return;
        #endif
    } else if (permission == PermissionGroupMicrophone) {
        #if PERMISSION_MICROPHONE
        mediaType = AVMediaTypeAudio;
        #else
        completionHandler(PermissionStatusUnknown);
        return;
        #endif
    } else {
        completionHandler(PermissionStatusUnknown);
        return;
    }

    [AVCaptureDevice requestAccessForMediaType:mediaType completionHandler:^(BOOL granted) {
        if (granted) {
            completionHandler(PermissionStatusGranted);
        } else {
            completionHandler(PermissionStatusDenied);
        }
    }];
}

+ (PermissionStatus)permissionStatus:(AVMediaType const)mediaType {
    AVAuthorizationStatus status = [AVCaptureDevice authorizationStatusForMediaType:mediaType];

    switch (status) {
        case AVAuthorizationStatusNotDetermined:
            return PermissionStatusUnknown;
        case AVAuthorizationStatusRestricted:
            return PermissionStatusRestricted;
        case AVAuthorizationStatusDenied:
            return PermissionStatusDenied;
        case AVAuthorizationStatusAuthorized:
            return PermissionStatusGranted;
    }

    return PermissionStatusUnknown;
}

@end

#else

@implementation AudioVideoPermissionStrategy
@end

#endif

File: ios/Classes/strategies/AudioVideoPermissionStrategy.swift
//
//  AudioVideoPermissions.swift
//  permission_handler
//
//  Created by Maurits van Beusekom on 26/07/2018.
//
import AVFoundation
import Foundation

class AudioVideoPermissionStrategy : NSObject, PermissionStrategy {
    
    func checkPermissionStatus(permission: PermissionGroup) -> PermissionStatus {
        if permission == PermissionGroup.camera {
            return AudioVideoPermissionStrategy.getPermissionStatus(mediaType: AVMediaType.video)
        } else if permission == PermissionGroup.microphone {
            return AudioVideoPermissionStrategy.getPermissionStatus(mediaType: AVMediaType.audio)
        }
        
        return PermissionStatus.unknown
    }
    
    private static func getPermissionStatus(mediaType: AVMediaType) -> PermissionStatus {
        let status: AVAuthorizationStatus = AVCaptureDevice.authorizationStatus(for: mediaType)
        
        switch status {
        case AVAuthorizationStatus.authorized:
            return PermissionStatus.granted
        case AVAuthorizationStatus.denied:
            return PermissionStatus.denied
        case AVAuthorizationStatus.restricted:
            return PermissionStatus.restricted
        default:
            return PermissionStatus.unknown
        }
    }
    
    func requestPermission(permission: PermissionGroup, completionHandler: @escaping PermissionStatusHandler) {
        let permissionStatus = checkPermissionStatus(permission: permission)
        
        if permissionStatus != PermissionStatus.unknown {
            completionHandler(permissionStatus)
            return
        }
        
        var mediaType: AVMediaType
        
        if permission == PermissionGroup.camera {
            mediaType = AVMediaType.video
        } else if permission == PermissionGroup.microphone {
            mediaType = AVMediaType.audio
        } else {
            completionHandler(PermissionStatus.unknown)
            return
        }
        
        AVCaptureDevice.requestAccess(for: mediaType, completionHandler: {
            (granted: Bool) in
                if granted {
                    completionHandler(PermissionStatus.granted)
                } else {
                    completionHandler(PermissionStatus.denied)
                }
        })
    }
}
